runtime/sam/expr/udf.go

```diff
@@ -1,43 +1,43 @@
 package expr
 
 import (
-	"slices"
-
 	"github.com/brimdata/super"
+	"github.com/brimdata/super/zcode"
 )
 
 const maxStackDepth = 10_000
 
 type UDF struct {
-	Body Evaluator
-	Name string
-	Sctx *super.Context
+	Body       Evaluator
+	sctx       *super.Context
+	name       string
+	fields     []super.Field
+	stackDepth *int
+	builder    zcode.Builder
 }
 
-func (u *UDF) Call(ectx super.Allocator, args []super.Value) super.Value {
-	stack := 1
-	if f, ok := ectx.(*frame); ok {
-		stack += f.stack
-	}
-	if stack > maxStackDepth {
-		return u.Sctx.NewErrorf("stack overflow in function %q", u.Name)
+func NewUDF(sctx *super.Context, name string, params []string, stackDepth *int) *UDF {
+	var fields []super.Field
+	for _, p := range params {
+		fields = append(fields, super.Field{Name: p})
 	}
-	// args must be cloned otherwise the values will be overwritten in
-	// recursive calls.
-	f := &frame{stack: stack, vars: slices.Clone(args)}
-	defer f.exit()
-	return u.Body.Eval(f, super.Null)
+	return &UDF{sctx: sctx, name: name, fields: fields, stackDepth: stackDepth}
 }
 
-type frame struct {
-	stack int
-	vars  []super.Value
-}
-
-func (f *frame) Vars() []super.Value {
-	return f.vars
-}
-
-func (f *frame) exit() {
-	f.stack--
+func (u *UDF) Call(ectx super.Allocator, args []super.Value) super.Value {
+	*u.stackDepth++
+	if *u.stackDepth > maxStackDepth {
+		return u.sctx.NewErrorf("stack overflow in function %q", u.name)
+	}
+	defer func() { *u.stackDepth-- }()
+	if len(args) == 0 {
+		return u.Body.Eval(NewContext(), super.Null)
+	}
+	u.builder.Reset()
+	for i := range args {
+		u.fields[i].Type = args[i].Type()
+		u.builder.Append(args[i].Bytes())
```

Comment:
Nit: Maybe make this a little more concise?

```suggestion
	for i, a := range args {
		u.fields[i].Type = a.Type()
		u.builder.Append(a.Bytes())
```